### ISSUE TITLE
x/build: add MSAN and ASAN builders

### ISSUE THREAD

**Issue by mknyszek** [golang/go#70054]
I recently broke ASAN mode because the tests in `cmd/cgo/internal/testsanitizers` are too simple and can't catch issues with Go's instrumentation of user code or issues with the instrumentation of the Go runtime. In this particular case, it was the latter: some allocation codepaths were broken under ASAN (did not notify ASAN of changes correctly), but more sophisticated allocation patterns were required to trigger the issue.

I propose we set up ASAN and MSAN builders, akin to race mode builders, that run all Go tests under MSAN and ASAN. At the very least, this will catch problems with how Go hooks into the sanitizer runtimes. I suspect these weren't set up in the past because ASAN/MSAN do little to help catch issues in Go code (not much opportunity for a use-after-free with a garbage collector, nor is uninitialized memory access likely to occur in regular Go code (I'm not even sure if we instrument Go code for the latter at all)). As long as we have spare machine resources, I think it's worth it to try and catch these lower-level issues early. However, since they're not _super_ high value builders, I think running them only in postsubmit is sufficient.

**Comment by mauri870** [comment]
Currently there are many tests that fail with ASAN/MSAN enabled, I looked a bit into this in separate issues. It seems that most of the failures are related to the number of allocations.

- https://github.com/golang/go/issues/64257
- https://github.com/golang/go/issues/64256

**Comment by mknyszek** [comment]
Yeah, I expected some issues. The alloc numbers are probably because of when the instrumentation happens (before inlining?) and so some things don't get inlined as expected, and we get some more allocs. (Correct me if I'm wrong, anyone.)

I don't think fixing the allocs is worth it; we should just skip those tests. There are other tests we should just skip as well. I'll send CLs today.

**Comment by mknyszek** [comment]
Okay, drilling down on one of the tests, `TestAppendOfMake`, reveals at least part of the problem: the instrumentation seems to defeat an optimization to eliminate `runtime.makeslice`. This doesn't happen with `race` mode, so there must be some handling for it already.

**Comment by mknyszek** [comment]
Just kidding, I totally missed that the test is skipped entirely if `!race`. Great, that makes this much easier -- no messing around with compiler optimizations, just skip.

**Comment by mauri870** [comment]
Thanks for investigating! I believe I already have a branch with changes to skip the tests, but there are so many skips that, at some point, it felt like I wasn’t doing things right.

I can open a CL with the changes. Let me know if you want to send one yourself.

**Comment by mauri870** [comment]
That is one of the reasons we introduced internal packages for mean/asan as well https://github.com/golang/go/issues/64611

**Comment by mknyszek** [comment]
Thanks. I missed your message; I just sent a CL with (almost) all the skips.

A good chunk of the skips do align with `-race` so I'm not too worried about it. It might be "better" in some sense to make `-asan` and `-msan` not defeat quite so many optimizations in certain specific cases, but my immediate goal is to just stop breaking them. I can file a new issue to track any places where, for example, `-asan` falls behind other modes (that definitely seems like the outlier).

**Comment by mknyszek** [comment]
OK, so the vast majority of test changes are skips, but there were some real bugs with ASAN that just weren't caught because we weren't running all of our tests. I've fixed a bunch of them, but there may be more to fix. Good progress so far, definitely much cleaner than when I started.

**Comment by mknyszek** [comment]
On a whim I just tried skipping `TestMemmoveOverflow` since it looked fishy and always seemed to be running in parallel with failing test (`=== CONT` in the output). Looks like that's the culprit. The runtime tests with `stress2` don't fail if I skip it.

**Comment by dmitshur** [comment]
Thanks for adding the builders and fixing tests at tip. The builders are currently failing on release branches 1.23 and 1.22: (CC @cagedmantis)

https://ci.chromium.org/p/golang/g/go-go1.23/console
https://ci.chromium.org/p/golang/g/go-go1.22/console

@mknyszek Do you think it's viable and worthwhile to try to fix the release branches too? Or is it better to just constrain these new builders to only run at 1.24 branch and newer? Making these new builders 1.24+ only seems fine to me, but up to you.

**Comment by mknyszek** [comment]
Hm... the benefit of backporting is to make sure that backports don't break MSAN/ASAN. But I think such a situation is quite rare. I also think it would be fine to make them 1.24+ only, though if it turns out the backports are _easy_. I can try to just backport the changes naively and see what happens.
